### PR TITLE
feat: AWS Org Support

### LINKED ISSUE #442: Support Fetching Via AWS Organizations
We support fetching data about Organizations, but we do not support traversing an OU structure to fetch data.

A customer should be able to specify An OU path and  A Role name and we should traverse the OU structure fetching data using the provided role.

caveat:
- The initial credentials must belong to an admin account

### PR DESCRIPTION
**Pull request by bbernays**
- [x] Relates OR Closes #442 
- [ ] Added integration tests as described [here](https://docs.cloudquery.io/docs/developers/sdk/testing)

Output from integration tests:
More information about running the tests [here](../docs/contributing/e2e_tests.md)

```
$ go test -tags=integration ./...

```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -343,16 +346,125 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
...
+	}
+	for _, account := range rawAccounts {
+		logger.Info("adding account", "account", account)
+		// Only load Active accounts
+		if account.Status != orgTypes.AccountStatusActive {
+			continue
+		}
+		accounts = append(accounts, Account{
+			ID:              *account.Id,
+			RoleARN:         fmt.Sprintf("arn:aws:iam::%s:role/%s", *account.Id, awsConfig.Organization.ChildAccountRoleName),
```
Comment (disq) [inline client/client.go:438]: Use `client.GenerateResourceARN`

File: client/client.go [hunk 2]
```diff
@@ -343,16 +346,125 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
...
+			LocalProfile:    awsConfig.Organization.AdminAccount.LocalProfile,
+			Regions:         awsConfig.Organization.ChildAccountRegions,
+			source:          "org",
+		})
+	}
+	return accounts, sts.NewFromConfig(awsCfg), err
+
+}
+
+func (org AwsOrg) Validate() bool {
```
Comment (disq) [inline client/client.go:450]: `IsValid`? If not valid, should we explicitly return an error? So maybe return an `error` to pass on the message? Not sure

**Review by disq — APPROVED** [review]
LGTM overall with two comments

File: client/client.go (hunk 2 above)
Comment (bbernays) [reply client/client.go:450]: I just decided to remove the entire function

File: client/client.go [hunk 3]
```diff
@@ -343,16 +346,132 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
...
+func loadOrgAccounts(ctx context.Context, logger hclog.Logger, awsConfig *Config) ([]Account, AssumeRoleAPIClient, error) {
+
+	adminAccount := Account{AccountName: "Default-Admin-Account"}
+	if awsConfig.Organization.AdminAccount != nil {
+		adminAccount = *awsConfig.Organization.AdminAccount
+	}
+	awsCfg, err := configureAwsClient(ctx, logger, awsConfig, adminAccount, nil)
+	if err != nil {
+		return nil, nil, err
+	}
```
Comment (roneli) [inline client/client.go:451]: document this line, as it is hard to follow the logic here, took me a few minutes to fully understand ~ what is done here.

File: client/client.go [hunk 4]
```diff
@@ -343,16 +346,132 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
...
+	svc := organizations.NewFromConfig(awsCfg)
+	accounts, err := loadAccounts(ctx, awsConfig, svc)
+	if err != nil {
+		return nil, nil, err
+	}
+	return accounts, sts.NewFromConfig(awsCfg), err
```
Comment (roneli) [inline client/client.go:457]: nit:
```suggestion
	return accounts, sts.NewFromConfig(awsCfg), nil
```

**Review by roneli — CHANGES_REQUESTED** [review]
Looks good! amazing feature added to AWS provider. had some nits and requests but looks good overall.

File: docs/index.md [hunk 5]
```diff
@@ -60,6 +60,97 @@ By default, CloudQuery will fetch all configuration from **all** supported resou
...
+  `organizations:ListChildren`
+
+2. Have a role in each child account that has a trust policy with the admin accounts. The default profile name is `OrganizationAccountAccessRole`
+
+
+Using AWS Organization:
+1. Specify member role name:
+
+```hcl
+    org {
```
[the remark this answers is not in the document]
Comment (bbernays) [reply docs/index.md:81]: I thought about that. I figured if you needed more than 1 org you can duplicate the provider

File: client/client.go [hunk 6]
```diff
@@ -343,16 +346,132 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
...
 
 	return awsCfg, err
 }
 
+type ListAccountsAPI interface {
+	ListAccountsForParent(ctx context.Context, params *organizations.ListAccountsForParentInput, optFns ...func(*organizations.Options)) (*organizations.ListAccountsForParentOutput, error)
+	ListAccounts(ctx context.Context, params *organizations.ListAccountsInput, optFns ...func(*organizations.Options)) (*organizations.ListAccountsOutput, error)
+}
+
+func getOUAccounts(ctx context.Context, accountsApi ListAccountsAPI, ous []string) ([]orgTypes.Account, error) {
```
[the remark this answers is not in the document]
Comment (bbernays) [reply client/client.go:361]: Not sure I understand. Like create a new organization.go file in the client package?

File: docs/index.md [hunk 7]
```diff
@@ -60,6 +60,97 @@ By default, CloudQuery will fetch all configuration from **all** supported resou
...
+
+CloudQuery supports discovery of AWS Accounts via AWS Organizations. This means that as Accounts get added or removed from your organization CloudQuery will be able to handle new or removed accounts without any configuration changes.
+
+Prerequisites for using AWS Org functionality:
+1. Have a role (or user) in an Admin account with the following access 
+  `organizations:ListAccounts`
+  `organizations:ListAccountsForParent`
+  `organizations:ListChildren`
+
+2. Have a role in each child account that has a trust policy with the admin accounts. The default profile name is `OrganizationAccountAccessRole`
```
[the remark this answers is not in the document]
Comment (bbernays) [reply docs/index.md]: Done

File: client/client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (roneli) [reply client/client.go]: yes

File: docs/index.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (roneli) [reply docs/index.md:81]: I think we should limit this then and fail with error if we have more than 1 block.

**Review by yevgenypats — APPROVED** [review]

**Review by roneli — APPROVED** [review]
LGTM, would maybe enforce 1 org block in configure, but other than that we can merge.
